File: problem_014.rs
// -*- compile-command: "rustc -o problem_014_rs problem_014.rs" -*-
// Copyright (c) 2016 Michael Caldwell


fn main() {
    println!("Project Euler - Problem 14");
    println!("Which starting number, under one million, produces the longest chain by following the rules below?\n\nif n is even, n/2\nif n is odd, 3n+1\n");

    let mut longest = [0, 0];
    let mut chainlength = Vec::new();
    chainlength.push(0);

    for n in 1..1000000 {
        let mut count = 0;
        let mut num = n;

        // Follow chain to 1
        while num != 1 {
            count += 1;
            if num % 2 == 0 {
                num /= 2;
            } else {
                num = num * 3 + 1;
            }
            if num < chainlength.len() {
                count += chainlength[num];
                num = 1;
            }
        }

        // Keep track of longest chain
        if count > longest[0] {
            longest[0] = count;
            longest[1] = n;
        }

        // Record this chain's length
        chainlength.push(count);
    }

    println!("{}", longest[1]);
}

File: problem_014.go
// -*- compile-command: "go build -o problem_014_go problem_014.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func main() {
        fmt.Println("Project Euler - Problem 14:")
        fmt.Println("Which starting number, under one million, produces the longest chain by following the rules below?\n")
	fmt.Println("if n is even, n/2")
	fmt.Println("if n is odd, 3n+1\n")


	limit := 1000000
	history := make([]int, limit)
	longest := 0

	for n := 2 ; n < limit; n++ {
		chain := 0
		number := n
		for number > 1 {
			// Follow rules
			if number % 2 == 0 {
				number /= 2
			} else {
				number = number * 3 + 1
			}
			chain++

			//check history to see if we can end early
			if number < limit && history[number] != 0 {
				chain += history[number]
				number = 1
			}
		}


		// Record history
		history[n] = chain

		// Check for new winner
		if chain > history[longest] {
			longest = n
		}
	}

	fmt.Println(longest)
}
